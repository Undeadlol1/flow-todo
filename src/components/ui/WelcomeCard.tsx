import React from 'react';
import Paper from '@material-ui/core/Paper';
import Typography from '@material-ui/core/Typography';
import { makeStyles } from '@material-ui/styles';
import Grid from '@material-ui/core/Grid';
import { Theme } from '@material-ui/core';

const useStyles = makeStyles((theme: Theme) => ({
  container: {
    padding: theme.spacing(3),
  },
  image: {
    width: '100%',
    margin: '0 auto',
    display: 'block',
    paddingBottom: theme.spacing(3),
  },
}));

const WelcomeCard: React.FC<{}> = () => {
  const classes = useStyles();
  return (
    <Paper className={classes.container} elevation={6}>
      <Grid container justify="center">
        <Grid item xs={7} sm={6} md={6} lg={6}>
          <img
            className={classes.image}
            src="/images/logo.png"
            alt=""
          />
        </Grid>
      </Grid>
      {/* TODO: i18n */}
      <Typography paragraph>
        Это приложение поможет справиться с задачами, которые ты вечно
        откладываешь в
        {' '}
        <b>долгий ящик</b>
        .
      </Typography>
      <Typography paragraph>
        Просто добавь пару задач и смотри, как приложение с помощью
        психологических трюков облегчает их и позволяет сдвинуться с
        мертвой точки.
      </Typography>
    </Paper>
  );
};

export default WelcomeCard;
