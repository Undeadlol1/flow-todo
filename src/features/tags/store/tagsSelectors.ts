import { RootReducer } from '../../../store';

export const activeTagsSelector = (state: RootReducer) =>
  state.tags.activeTags;
