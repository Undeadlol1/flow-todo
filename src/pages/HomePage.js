import React from 'react';
import Grid from '@material-ui/core/Grid';
import Fab from '@material-ui/core/Fab';
import AddIcon from '@material-ui/icons/Add';
import { makeStyles } from '@material-ui/core/styles';
import Dialog from '@material-ui/core/Dialog';
import DialogContent from '@material-ui/core/DialogContent';
import { useTranslation } from 'react-i18next';
import TasksList from '../components/tasks/TasksList/TasksList';
import RandomTaskButton from '../components/tasks/RandomTaskButton/RandomTaskButton';
import CreateTask from '../components/tasks/CreateTask/CreateTask';

const useStyles = makeStyles(theme => ({
  pageContainer: {
    minHeight: 'calc(100vh - 64px)',
  },
  fab: {
    position: 'fixed',
    [theme.breakpoints.down('sm')]: {
      bottom: theme.spacing(2),
      right: theme.spacing(2),
    },
    [theme.breakpoints.up('md')]: {
      bottom: theme.spacing(4),
      right: theme.spacing(4),
    },
    [theme.breakpoints.up('lg')]: {
      bottom: theme.spacing(6),
      right: theme.spacing(8),
    },
  },
}));

export default function HomePage() {
  const classes = useStyles();
  const [t] = useTranslation();
  const [isDialogOpen, setDialog] = React.useState(false);
  const toggleDialog = () => setDialog(!isDialogOpen);
  return (
    <Grid
      container
      spacing={2}
      justify="center"
      direction="column"
      alignItems="stretch"
      alignContent="center"
      className={classes.pageContainer}
    >
      <Grid item xs={12} sm={8} md={8} lg={6} align="center">
        <RandomTaskButton />
      </Grid>
      <Grid item xs={12} sm={8} md={8} lg={6} align="center">
        <TasksList />
      </Grid>
      <Dialog
        open={isDialogOpen}
        aria-labelledby="form-dialog-title"
        onClose={toggleDialog}
      >
        <DialogContent>
          <CreateTask />
        </DialogContent>
      </Dialog>
      <Fab
        color="primary"
        aria-label={t('createTask')}
        className={classes.fab}
        onClick={toggleDialog}
      >
        <AddIcon />
      </Fab>
    </Grid>
  );
}
