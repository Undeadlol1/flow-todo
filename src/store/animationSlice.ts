import { createSlice, PayloadAction } from '@reduxjs/toolkit';

export type AnimationState = {
  isPointsRewardingInProgress: boolean;
  pointToDisplayDuringRewardAnimation: number;
};

const initialState: AnimationState = {
  isPointsRewardingInProgress: false,
  pointToDisplayDuringRewardAnimation: 0,
};

const animationSlice = createSlice({
  initialState,
  name: 'animation',
  reducers: {
    startPointsRewardingAnimation(
      state,
      { payload }: PayloadAction<number>,
    ) {
      state.isPointsRewardingInProgress = true;
      state.pointToDisplayDuringRewardAnimation = payload;
    },
    stopPointsRewardingAnimation(state) {
      state.isPointsRewardingInProgress = false;
    },
  },
});

export const {
  stopPointsRewardingAnimation,
  startPointsRewardingAnimation,
} = animationSlice.actions;

export default animationSlice.reducer;
