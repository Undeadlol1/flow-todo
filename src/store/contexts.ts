import { createContext } from 'react';
import { firestore } from 'firebase';

type ContextProps = {
  loading: boolean;
  tasks: firestore.QuerySnapshot;
  tasksDoneToday: firestore.QuerySnapshot;
};

export const TasksContext = createContext({
  tasks: {},
  loading: true,
  tasksDoneToday: {},
} as Partial<ContextProps>);
