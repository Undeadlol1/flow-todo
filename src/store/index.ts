import nanoid from 'nanoid';
import { firestore } from 'firebase/app';
import subtractDays from 'date-fns/subDays';
import {
  configureStore,
  getDefaultMiddleware,
  combineReducers,
} from '@reduxjs/toolkit';
import tasksSlice from './tasksSlice';
import uiSlice from './uiSlice';
import { useSelector, TypedUseSelectorHook } from 'react-redux';
import extend from 'lodash/extend';

export type Task = {
  name: string;
  isDone: boolean;
  doneAt?: number;
  dueAt: number;
  userId: string;
  note?: string;
  subtasks?: any[];
};

export function upsertTask(
  values: { name?: string; userId?: string; isCurrent?: boolean },
  taskId?: string,
): Promise<void | Error> {
  const isCreate = !taskId;
  const payload = extend(
    values,
    isCreate && {
      isDone: false,
      dueAt: subtractDays(new Date(), 1).getTime(),
    },
  );

  if (isCreate && !values.userId)
    return Promise.reject('You forgot to add userId');

  return firestore()
    .collection('tasks')
    .doc(taskId || nanoid())
    .set(payload, { merge: true });
}

export function deleteTask(taskId: string): Promise<void | Error> {
  return firestore()
    .doc('tasks/' + taskId)
    .delete();
}

export function createSubtask(
  taskId: string,
  values: {
    name: string;
  },
): Promise<void | Error> {
  return firestore()
    .doc('tasks/' + taskId)
    .update({
      subtasks: firestore.FieldValue.arrayUnion({
        id: nanoid(),
        isDone: false,
        parentId: taskId,
        createdAt: Date.now(),
        name: values.name.trim(),
      }),
    });
}

export type Subtask = {
  id: string;
  isDone: boolean;
  parentId: string;
  createdAt: number;
  name: string;
};

export async function updateSubtask(
  subtask: Subtask,
  values: {
    name?: string;
    doneAt: number;
    isDone: boolean;
  },
): Promise<void | Error> {
  const docRef = firestore().doc('tasks/' + subtask.parentId);
  const task: any = await docRef.get();
  const newSubtasks: any[] = task
    .data()
    .subtasks.map((i: Subtask) => {
      return i.id === subtask.id ? Object.assign({}, i, values) : i;
    });
  return docRef.update({
    subtasks: newSubtasks,
  });
}

export function deleteSubtask(
  taskId: string,
  subtask: {
    id: string;
  },
): Promise<void | Error> {
  return firestore()
    .doc('tasks/' + taskId)
    .update({
      subtasks: firestore.FieldValue.arrayRemove(subtask),
    });
}

const rootReducer = combineReducers({
  ui: uiSlice,
  tasks: tasksSlice,
});

const store = configureStore({
  reducer: rootReducer,
  middleware: [...getDefaultMiddleware()],
  devTools: process.env.NODE_ENV !== 'production',
  enhancers: [],
});

export const useTypedSelector: TypedUseSelectorHook<
  ReturnType<typeof rootReducer>
> = useSelector;

export default store;
// The store has been created with these options:
// - The slice reducers were automatically passed to combineReducers()
// - redux-thunk and redux-logger were added as middleware
// - The Redux DevTools Extension is disabled for production
// - The middleware, batch, and devtools enhancers were automatically composed together
