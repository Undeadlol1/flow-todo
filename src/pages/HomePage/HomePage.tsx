import Grid from '@material-ui/core/Grid';
import { makeStyles } from '@material-ui/core/styles';
import debug from 'debug';
import isEmpty from 'lodash/isEmpty';
import isUndefined from 'lodash/isUndefined';
import React, { memo } from 'react';
import CreateTaskFab from '../../components/tasks/CreateTaskFab';
import GetRandomTask from '../../components/tasks/RandomTaskButton/RandomTaskButton';
import AppTour from '../../components/ui/AppTour';
import WelcomeCard from '../../components/ui/WelcomeCard';
import { Task, useTypedSelector } from '../../store/index';
import PinnedTask from '../../components/tasks/TasksList/PinnedTask';
import { tasksSelector, uiSelector } from '../../store/selectors';
import get from 'lodash/fp/get';
import clsx from 'clsx';
import { useScreenIsNarrow } from '../../services/index';
import { TagsList } from '../../components/tasks/TagsList';
import TasksDoneToday from '../../components/tasks/TasksDoneToday';
import { Unless } from 'react-if';

const log = debug('HomePage');
const useStyles = makeStyles(theme => ({
  pageContainer: {
    marginTop: 0,
    marginBottom: 0,
    minHeight: 'calc(100vh - 74px)',
  },
  randomButtonContainer: {
    margin: '0 auto',
  },
  fullWidth: {
    width: '100%',
  },
}));

interface Props {
  activeTasks?: Task[];
  isLoading?: boolean;
  isAppTourActive?: boolean;
  createdAtleastOneTask?: Task[];
}

export const HomePage = memo(function HomePage(props: Props) {
  const classes = useStyles();
  const isScreeenNarrow = useScreenIsNarrow();

  const {
    isLoading,
    activeTasks,
    isAppTourActive,
    createdAtleastOneTask,
  } = props;
  log('isLoading: ', isLoading);
  log('activeTasks: ', activeTasks);
  log('isAppTourActive: ', isAppTourActive);
  log('createdAtleastOneTask: ', createdAtleastOneTask);

  function renderButtonOrWelcomeCard() {
    if (
      isLoading ||
      isAppTourActive ||
      !isEmpty(createdAtleastOneTask)
    )
      return <GetRandomTask className={'IntroHandle__taskButton'} />;
    else return <WelcomeCard />;
  }

  return (
    <Grid
      container
      spacing={2}
      justify="center"
      direction="column"
      alignItems="stretch"
      alignContent="center"
      className={classes.pageContainer}
    >
      <Grid item xs={12} sm={12} md={8} lg={6}>
        <Unless
          condition={Boolean(
            isEmpty(createdAtleastOneTask) || isAppTourActive,
          )}
        >
          <TasksDoneToday />
        </Unless>
      </Grid>
      <Grid item xs={12} sm={12} md={8} lg={6}>
        <PinnedTask />
      </Grid>
      <Grid
        item
        xs={12}
        sm={12}
        md={8}
        lg={6}
        className={clsx(
          classes.randomButtonContainer,
          isScreeenNarrow && classes.fullWidth,
        )}
      >
        {renderButtonOrWelcomeCard()}
      </Grid>
      <Grid item xs={12} sm={8} md={8} lg={6}>
        <TagsList />
      </Grid>
      <CreateTaskFab
        isHidden={isLoading}
        className="IntroHandle__createTask"
      />
      <AppTour />
    </Grid>
  );
});

export default memo(function HomePageContainer(props) {
  const activeTasks = useTypedSelector(tasksSelector);
  const { createdAtleastOneTask } = useTypedSelector(
    get('firestore.ordered'),
  );
  const { isAppTourActive } = useTypedSelector(uiSelector);
  const isLoading =
    isUndefined(createdAtleastOneTask) || isUndefined(activeTasks);
  return (
    <HomePage
      {...{
        isLoading,
        activeTasks,
        isAppTourActive,
        createdAtleastOneTask,
      }}
    />
  );
});
