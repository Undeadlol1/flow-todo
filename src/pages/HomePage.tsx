import Dialog from '@material-ui/core/Dialog';
import DialogContent from '@material-ui/core/DialogContent';
import Grid from '@material-ui/core/Grid';
import { makeStyles } from '@material-ui/core/styles';
import AddIcon from '@material-ui/icons/Add';
import cx from 'clsx';
import debug from 'debug';
import { auth } from 'firebase/app';
import React, { memo, useContext } from 'react';
import { useAuthState } from 'react-firebase-hooks/auth';
import { useTranslation } from 'react-i18next';
import useToggle from 'react-use-toggle';
import UpsertTask from '../components/tasks/CreateTask/UpsertTask';
import GetRandomTask from '../components/tasks/RandomTaskButton/RandomTaskButton';
import TasksList from '../components/tasks/TasksList/TasksList';
import AppTour from '../components/ui/AppTour';
import Fab from '../components/ui/Fab';
import WelcomeCard from '../components/ui/WelcomeCard';
import { TasksContext } from '../store/contexts';
import { useTypedSelector } from '../store/index';

const log = debug('HomePage');
debug.enable('HomePage');
const useStyles = makeStyles(theme => ({
  pageContainer: {
    minHeight: 'calc(100vh - 64px)',
  },
  randomButtonContainer: {
    margin: '0 auto',
  },
}));

export default memo(function HomePage() {
  const classes = useStyles();
  const [t] = useTranslation();
  const { loading: isLoading } = useContext(TasksContext);
  const { isAppTourActive } = useTypedSelector(state => state.ui);
  // @ts-ignore
  const { createdAtleastOneTask } = useTypedSelector(
    // @ts-ignore
    state => state.firestore.ordered,
  );

  log('createdAtleastOneTask: ', createdAtleastOneTask);
  const isLoggedIn = Boolean(useAuthState(auth())[0]);
  const [isDialogOpen, toggleDialog] = useToggle(false);
  const isButtonVisible =
    isLoading || isAppTourActive || !isEmpty(createdAtleastOneTask);

  return (
    <Grid
      container
      spacing={2}
      justify="center"
      direction="column"
      alignItems="stretch"
      alignContent="center"
      className={classes.pageContainer}
    >
      <Grid
        item
        xs={12}
        sm={8}
        md={8}
        lg={6}
        className={classes.randomButtonContainer}
      >
        {isButtonVisible && (
          <GetRandomTask className={'IntroHandle__taskButton'} />
        )}
        {!isButtonVisible && <WelcomeCard />}
      </Grid>
      <Grid item xs={12} sm={8} md={8} lg={6}>
        <TasksList />
      </Grid>
      <Dialog
        open={isDialogOpen}
        onClose={toggleDialog}
        aria-labelledby="form-dialog-title"
      >
        <DialogContent>
          <UpsertTask autoFocus beforeSubmitHook={toggleDialog} />
        </DialogContent>
      </Dialog>
      <Fab
        onClick={toggleDialog}
        aria-label={t('createTask')}
        className={cx(['IntroHandle__createTask'])}
        isHidden={isLoading || (!isLoggedIn && !isAppTourActive)}
      >
        <AddIcon />
      </Fab>
      <AppTour />
    </Grid>
  );
});
