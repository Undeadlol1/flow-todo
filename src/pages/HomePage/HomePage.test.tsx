import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';

import store from '../../store';
import HomePage from './HomePage';
import { BrowserRouter } from 'react-router-dom';
import WelcomeCard from './../../components/ui/WelcomeCard';
import { shallow } from 'enzyme';

describe('<HomePage />', () => {
  const defaultProps = {};
  const component = (
    <Provider store={store}>
      <BrowserRouter>
        <HomePage {...defaultProps} />
      </BrowserRouter>
    </Provider>
  );

  test('render', () => {
    const wrapper = renderer.create(component);
    expect(wrapper).toMatchSnapshot();
  });

  test('contains WelcomeCard', () => {
    const wrapper = shallow(component);
    console.log('debug', wrapper.debug());
  });
});
