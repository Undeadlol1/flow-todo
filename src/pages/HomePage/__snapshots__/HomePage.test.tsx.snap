// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomePage /> render 1`] = `
<div
  className="MuiGrid-root makeStyles-pageContainer-1 MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-direction-xs-column MuiGrid-align-content-xs-center MuiGrid-justify-xs-center"
>
  <div
    className="MuiGrid-root makeStyles-randomButtonContainer-2 MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-8 MuiGrid-grid-md-8 MuiGrid-grid-lg-6"
  >
    <div
      aria-disabled={true}
      className="MuiButtonBase-root MuiButton-root makeStyles-buttonRoot-107 MuiButton-text RandomTaskButton IntroHandle__taskButton MuiButton-textPrimary Mui-disabled Mui-disabled"
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex={-1}
      to="/tasks/undefined"
    >
      <span
        className="MuiButton-label"
      >
        <div
          className="MuiPaper-root MuiPaper-elevation6 makeStyles-paper-106 MuiPaper-rounded"
        >
          <div
            className="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
            role="progressbar"
            style={
              Object {
                "height": 40,
                "width": 40,
              }
            }
          >
            <svg
              className="MuiCircularProgress-svg"
              viewBox="22 22 44 44"
            >
              <circle
                className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
                cx={44}
                cy={44}
                fill="none"
                r={20.2}
                strokeWidth={3.6}
                style={Object {}}
              />
            </svg>
          </div>
        </div>
      </span>
    </div>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-8 MuiGrid-grid-md-8 MuiGrid-grid-lg-6"
  />
  <div
    className="react-joyride"
  />
</div>
`;
